v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/rowmapper/provider/PostgreSQLJdbcValueMappings.java

```diff
@@ -19,34 +19,216 @@
+          .optionalEnd()
+          .appendOffset("+HH:mm", "+00")
+          .toFormatter();
+
+  private static long toMicros(Instant instant) {
+    return TimeUnit.SECONDS.toMicros(instant.getEpochSecond())
+        + TimeUnit.NANOSECONDS.toMicros(instant.getNano());
+  }
+
+  private static long toMicros(OffsetTime offsetTime) {
+    return TimeUnit.HOURS.toMicros(offsetTime.getHour())
+        + TimeUnit.MINUTES.toMicros(offsetTime.getMinute())
+        + TimeUnit.SECONDS.toMicros(offsetTime.getSecond())
+        + TimeUnit.NANOSECONDS.toMicros(offsetTime.getNano());
+  }
+
   private static final ResultSetValueMapper<?> valuePassThrough = (value, schema) -> value;
 
-  private static final Calendar utcCalendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+  private static final ResultSetValueExtractor<ByteBuffer> bytesExtractor =
+      (rs, fieldName) -> {
+        byte[] bytes = rs.getBytes(fieldName);
+        if (bytes == null) {
+          return null;
+        }
+        return ByteBuffer.wrap(bytes);
+      };
 
-  private static final ResultSetValueExtractor<java.sql.Date> utcDateExtractor =
-      (rs, fieldName) -> rs.getDate(fieldName, utcCalendar);
+  private static final ResultSetValueExtractor<java.sql.Date> dateExtractor =
+      (rs, fieldName) -> rs.getDate(fieldName, UTC_CALENDAR);
 
-  private static final ResultSetValueMapper<java.sql.Date> sqlDateToAvro =
+  // We cannot use `java.sql.Time` here, as the PostgreSQL time might be '24:00:00'. This makes the
+  // java.sql.Time jump a day (from the underlying `java.sql.Date`) and microseconds extraction gets
+  // a wrong value. We parse the time from a String into an `OffsetTime` instead.
+  private static final ResultSetValueExtractor<OffsetTime> timeExtractor =
+      (rs, fieldName) -> {
+        String time = rs.getString(fieldName);
+        if (time == null) {
+          return null;
+        }
+        return OffsetTime.parse(time + ZoneOffset.UTC, TIME_FORMAT);
+      };
+
+  private static final ResultSetValueExtractor<OffsetTime> timetzExtractor =
+      (rs, fieldName) -> {
+        String timeTz = rs.getString(fieldName);
+        if (timeTz == null) {
+          return null;
+        }
+        return OffsetTime.parse(timeTz, TIMETZ_FORMAT);
+      };
+
+  private static final ResultSetValueExtractor<java.sql.Timestamp> timestampExtractor =
+      (rs, fieldName) -> rs.getTimestamp(fieldName, UTC_CALENDAR);
+
+  private static final ResultSetValueExtractor<OffsetDateTime> timestamptzExtractor =
+      (rs, fieldName) -> {
+        String timestampTz = rs.getString(fieldName);
+        if (timestampTz == null) {
+          return null;
+        }
+        return OffsetDateTime.parse(timestampTz, TIMESTAMPTZ_FORMAT);
+      };
+
+  // Value might be a Double.NaN or a valid BigDecimal
+  private static final ResultSetValueMapper<Number> numericToAvro =
+      (value, schema) -> value.toString();
```

Comment:
should we consider the precision and scale from the schema argument as required by the [unified-types](https://cloud.google.com/datastream/docs/unified-types) documentation?

Reply:
Yes, I would like to do that in a future PR, since this approach works with and without precision for now. I also have to do that for VARCHAR

Reply:
Added TODO